### DISCUSSION TITLE
Please share your feedback

### DISCUSSION THREAD

**Discussion by pthom** [pthom/imgui_bundle#74] [General]
"ImGui Bundle" is a brand new python and C++ package (Oct 2022). It works on MacOS, Windows and Linux.

Feedback from users would be appreciated in order to make sure its installation works on different platforms.

If you installed it, and if you have time, could you please answer the following:

_Python users_

- How did you install imgui-bundle for python: via `git clone` + `pip install -v .` or via `pip install imgui-bundle` (i.e install from pypi)
- If you installed from pypi, which platform are you using ? Did the build use the prebuilt wheel, or did it have to build from source?
- Does imgui_bundle_demo work ?
- Do you have code completion for imgui and implot in your IDE (and which IDE are you using)? If not, please read the [FAQ](https://github.com/pthom/imgui_bundle/blob/main/FAQ.md)

_C++ Users_
- Did the code compile painlessly, and do the demo work?

Thanks in advance!

**Comment by pthom** [comment]
Also, I'm looking for small educational examples such as [haiku_implot_heart.py](https://github.com/pthom/imgui_bundle/blob/main/bindings/imgui_bundle/demos/haikus/haiku_implot_heart.py) which would serve as a demonstration of the library. 
Please share your creations if you have small self contained demos which you think might be an help for new users 🙏.

**Comment by district10** [comment]
macOS m1, python user, install by `pip install imgui-bundle` (fails at first, succeed after `pip install cmake` (even I have cmake installed already).

`imgui_bundle_demo` works pretty well. Thank you for this nice work.

Feedbacks?

<img width="981" alt="image" src="https://user-images.githubusercontent.com/5262037/201331847-bfa9bbec-1f87-49e6-8411-e9fe4e87fc64.png">

and it deserves a better dock icon:

<img width="110" alt="image" src="https://user-images.githubusercontent.com/5262037/201331584-7b590c2a-b1ea-43f5-9a98-8cd882c6e27a.png">

and a 3d scene example will be great, (https://github.com/stevenlovegrove/Pangolin is hard to install.)
Maybe integrate https://github.com/CedricGuillemet/ImGuizmo?

I'm expecting to use this to try out some functions from https://github.com/libigl/libigl.

Thank you again.

**Comment by pthom** [comment]
@district10 : thanks for your message!

[v0.6.5](https://github.com/pthom/imgui_bundle/releases/tag/v0.6.5) is out. It solves the blurry text issue. 

Concerning 3D + ImGuizmo, I still need to ponder whether it is easy or not. I will study this in the next weeks.
Concerning the icon, I agree with you. However, bundling icons on a mac is cumbersome compared to windows so I had to postpone this.

**Comment by pthom** [comment]
> and a 3d scene example will be great, (https://github.com/stevenlovegrove/Pangolin is hard to install.)
> Maybe integrate https://github.com/CedricGuillemet/ImGuizmo?

v0.7.2 is out and integrates ImGuizmo (as well as lots of additions)!

**Comment by Rusty-Weasel** [comment]
Basically i like the idea to collect all cool imgui stuff  under one package, 
but really do not like the mixup between different languages and tool sets.

i would lovelier see a clear and cleaner structure and an easy fast way, something like following, then we get cool stuff:
![image](https://user-images.githubusercontent.com/61587283/211849955-48faff26-f672-4f91-bfbb-d85ce33cb104.png)

Further this example does not compile, as _RobotoRegularItalic.ttf_ is missing, but requested
![image](https://user-images.githubusercontent.com/61587283/211853298-b87e1178-02ed-41f2-be94-6ad54da777c8.png)

The python way is fast and simple, at c++ way i can just start debugging, 
but that was not, what i expected for a short and fast demo run

**Comment by pthom** [comment]
@GoWeasel : ...Hello and happy new year...

> but really do not like the mixup between different languages and tool sets

I did not understand what you mean, although I tried for a few minutes.
What mixup are you referring to ?
- Is it the mixup between python and C++ ? This repo provides C++ and python, because this project provides bindings for python (with a symetric API which makes me quite proud) ?
- Is is because you saw some python scripts that helps generate the pybind11 bindings?
- Is it anything else? 

Be precise, please.

>  a clear and cleaner structure

Well, thanks for the not so kind words with no clear explanation.

> `ImGuiBundle::ShowFullDemoWindow()`:

Why not, I might study this idea later. This is a good idea.

> The python way is fast and simple

Do you mean that you launched the python demos and they were simple to launch? In that case, thanks.

> Further this example does not compile, as RobotoRegularItalic.ttf is missing, but requested

Will be fixed.

> at c++ way i can just start debugging, but that was not, what i expected for a short and fast demo run

Does it mean you are asking for ImGuiBundle::ShowFullDemoWindow() or what?

In the meantime you can do :
```bash
mkdir build
cd build
cmake ..
./bin/demo_all
```

**Comment by Rusty-Weasel** [comment]
> > at c++ way i can just start debugging, but that was not, what i expected for a short and fast demo run

just maked an empty testfolder, then git clone (main) repository and open up as a cmake project, 
tada **cmake error saying hello to me**

![image](https://user-images.githubusercontent.com/61587283/211920668-e66c6f4e-614c-462d-afaa-e5a95b77c491.png)

What i would expect were something like, a demo.cpp which is build able with root cmake config, 

> In the meantime you can do :
> 
> ```shell
> mkdir build
> cd build
> cmake ..
> ./bin/demo_all
> ```

not possible with cmake errors

**Comment by Rusty-Weasel** [comment]
> Does it mean you are asking for ImGuiBundle::ShowFullDemoWindow() or what?

just expec to import ImGuiBundle like a lib (cmake compatible) that gives me all what i need (as so far implemented),
and then use it.

lets say for **project a** i just need a short imgui demo without big extras, i import ImGuiBundle Lib and just call 
either ImGuiBundle::ImGui::ShowDemoWindow() or just ImGui::ShowDemoWindow() (which already works)

now lets say for **project b** i like to see a full demo with all included imgui stuff, so i just call
ImGuiBundle::ShowFullDemoWindow()

now lets say for **project c**, i like to use different parts (of ImGuiBundle) and perhaps different gui-toolkit (glfw/sdl) and perhaps different graphic api (opengl/vulkan) and perhaps target OS (win/linux/mac/android/ios/web) etc.... (this to handle in a way like KISS-principle would be the biggest challenge)

for the other question u got tomorrow some examples

**Comment by pthom** [comment]
it seems you forgot a classic step with repositories which have submodules:
```
cd imgui_bundle
git submodule update --init --recursive
```

**Comment by pthom** [comment]
> different gui-toolkit (glfw/sdl)

It is already possible to target glfw or sdl, since ImGuiBundle is based on my other library HelloImGui (see its [api](https://github.com/pthom/hello_imgui/blob/master/src/hello_imgui/hello_imgui_api.md#api), and [runner_params.h](https://github.com/pthom/hello_imgui/blob/master/src/hello_imgui/runner_params.h)  which enables to select the backend between Sdl and glfw.

> perhaps target OS (win/linux/mac/android/ios/web) : 

ImguiBundle already targets windows, mac, ios and emscripten. 
See [live emscripten demo](https://traineq.org/ImGuiBundle/emscripten/bin/demo_imgui_bundle.html) 

Concerning iOS and android. [HelloImGui](https://github.com/pthom/hello_imgui) (on which ImGuiBundle is based) [also supports them](https://traineq.org/HelloImGui_6_Platforms.mp4). I did not make any effort in checking whether its is still the case when added to ImGuiBundle. I do not maintain the android platform anymore, since I do not own an android device, and the android toolset is too clunky for my taste.

**Comment by pthom** [comment]
@GoWeasel : please see updated instructions for c++ [here](https://pthom.github.io/imgui_bundle/#_usage_with_c)

**Comment by Rusty-Weasel** [comment]
> @GoWeasel : please see updated instructions for c++ [here](https://pthom.github.io/imgui_bundle/#_usage_with_c)

oh nice... looks pretty clean style... i'll take it as good night lecture :-)
btw got it so far running.

**Comment by KevinHQChen** [comment]
I'm interested in imgui_bundle to make my existing C++ research project compatible with multiple platforms.

The example C++ demo compiled pretty easily, but I have some build system questions.

~~I would like to include imgui_bundle into an external library, then link that library into my executable app. However it looks like the imgui_bundle_add_app function does not allow that? Is there a way to just create a library that includes imgui_bundle and defer the add_executable step?~~ never mind I just didn't understand FetchContent lol

I was also wondering if there is a way to pin versions of each submodule in imgui_bundle. I'm a novice when it comes to C++ build systems, but to me it looks like I'd have to build imgui_bundle from source each time I want to change the version of each submodule? Is there an easier way to do this through cmake?

**Comment by pthom** [comment]
@KevinHQChen 

> I'm interested in imgui_bundle to make my existing C++ research project compatible with multiple platforms.

Thanks for your input! Do not hesitate to share images of your project if yo feel like it. Which libraries do you intend to use.

imgui_bundle_add_app is just a variation on 

```cmake
add_executable( exe_name yourfiles.cpp)
target_link_libraries(exe_name imgui_bundle)
# plus copy assets/ so that they can easily be found
```

> I was also wondering if there is a way to pin versions of each submodule in imgui_bundle. I'm a novice when it comes to C++ build  systems, but to me it looks like I'd have to build imgui_bundle from source each time I want to change the version of each submodule? Is there an easier way to do this through cmake?

Two possibilities:

1/ If you are using the [CMakeLists](https://github.com/pthom/imgui_bundle/blob/main/_example_integration/CMakeLists.txt) from example_integration, you can set a fixed commit hash or tag beside GIT_TAG
```
FetchContent_Declare(
    imgui_bundle
    GIT_REPOSITORY https://github.com/pthom/imgui_bundle.git
    GIT_PROGRESS TRUE
    # Enter the desired git tag below
    GIT_TAG YourDesiredHash_Or_Tag
)
```

2/ Or you can use imgui_bundle as a submodule.

In either case, imgui_bundle version  will be fixed, and that will pin the version of the submodules it uses. 

If you want to change the version of the submodules independently from imgui_bundle main repo, you will have to fork it.

**Comment by corbane** [comment]
Hello, Already, a big Merci for this project which allows me to learn a lot about python+imgui+openCV!!

I'm using `pip install imgui-bundle`, I'm on Windows and `imgui.show_demo_window()` works fine.

I didn't understand the "demo_docking" example.
After some testing, it seems to me that using `DockableWindow` and `DockingSplit` should be part of an advanced usage demo.
Using `ImGuiWindowParams.default_imgui_window_type=DefaultImGuiWindowType.provide_full_screen_dock_space` with `AppWindowParams.restore_previous_geometry=True` is enough to get a demo of the Docking API.

![demo_simple_docking](https://user-images.githubusercontent.com/8430545/236685787-b6a10415-f97a-4efb-9f80-59864cf0ac1e.png)

```py
from dataclasses import dataclass, field
from imgui_bundle import imgui
from imgui_bundle import hello_imgui
from imgui_bundle import immapp
from imgui_bundle import implot 


@dataclass
class SubWindow:

    namedId: str
    isShow: bool = True

    def drawUi (self):
        if not self.isShow: return
        imgui.begin (self.namedId)
        if imgui.button ("Close"):
            self.isShow = False
        imgui.end ()


@dataclass
class App:
    
    sub_windows: list[SubWindow] = field (default_factory=list)
    show_demo_window: bool = False

    def start (self):
        runner_params = hello_imgui.RunnerParams (
            app_window_params = hello_imgui.AppWindowParams (
                window_title = 'Demo Simple Dockable Windows',
                restore_previous_geometry = True
            ),
            imgui_window_params = hello_imgui.ImGuiWindowParams (
                default_imgui_window_type = hello_imgui.DefaultImGuiWindowType.provide_full_screen_dock_space
            ),
            callbacks = hello_imgui.RunnerCallbacks (
                show_gui = self.drawUi
            )
        )
        immapp.run (runner_params)

    def drawUi (self):
        isOpen, _ = imgui.begin ('Main Window')
        if isOpen:
           if imgui.button ('Create new window'):
                self.sub_windows.append (SubWindow (f'Sub Window {len (self.sub_windows)}'))
            if imgui.button ('Hide demo window' if self.show_demo_window else 'Show demo window'):
                self.show_demo_window = not self.show_demo_window
        imgui.end ()

        for win in self.sub_windows:
            win.drawUi ()

        if self.show_demo_window:
            imgui.show_demo_window ()


if __name__ == '__main__':
    App ().start ()
```

**Comment by pthom** [comment]
@corbane  : your example is OK. However the intent of the docking demo is to show how to handle the layout programmatically. 

I'll keep in mind that a simpler demo might be useful.

**Comment by learn-more** [comment]
Hello,

From poking around in the documentation this seems like a very nice project!

However, I was unable to find an option to disable certain 'batteries included' libraries (With the exception of `ImmVision`.
For example, I would like to disable `imgui_tex_inspect`, `ImFileDialog`, `ImGuizmo`, `ImPlot`, `imgui-node-editor`.

Is this a feature you would consider adding?

**Reply by pthom** [reply]
Hello, many thanks for your feedback! 

> However, I was unable to find an option to disable certain 'batteries included' libraries (With the exception of ImmVision.
For example, I would like to disable imgui_tex_inspect, ImFileDialog, ImGuizmo, ImPlot, imgui-node-editor.

In python, no. In C++, maybe. 

Are you talking about python or C++?
If targeting C++ what is the reason for asking, did you experience compilation issues; or do you simply want to reduce the binary size, the compilation time, etc?

**Reply by learn-more** [reply]
> Are you talking about python or C++? If targeting C++ what is the reason for asking, did you experience compilation issues; or do you simply want to reduce the binary size, the compilation time, etc?

Sorry, talking about c++.

The goal is indeed trying to reduce binary size, compilation time and the likes.

**Reply by pthom** [reply]
You can try the branch https://github.com/pthom/imgui_bundle/compare/disable_libs?expand=1

**Reply by learn-more** [reply]
> You can try the branch https://github.com/pthom/imgui_bundle/compare/disable_libs?expand=1

Wow, that was fast!
That appears to do exactly what I want, thanks!
